spring-core/src/main/java/org/springframework/core/GenericTypeResolver.java

```diff
@@ -167,25 +168,32 @@ public static Type resolveType(Type genericType, @Nullable Class<?> contextClass
 			}
 			else if (genericType instanceof ParameterizedType parameterizedType) {
 				ResolvableType resolvedType = ResolvableType.forType(genericType);
-				Class<?>[] generics = new Class<?>[parameterizedType.getActualTypeArguments().length];
+				ResolvableType[] generics = new ResolvableType[parameterizedType.getActualTypeArguments().length];
 				Type[] typeArguments = parameterizedType.getActualTypeArguments();
 				ResolvableType contextType = ResolvableType.forClass(contextClass);
 				for (int i = 0; i < typeArguments.length; i++) {
 					Type typeArgument = typeArguments[i];
 					if (typeArgument instanceof TypeVariable<?> typeVariable) {
 						ResolvableType resolvedTypeArgument = resolveVariable(typeVariable, contextType);
 						if (resolvedTypeArgument != ResolvableType.NONE) {
-							generics[i] = resolvedTypeArgument.resolve();
+							generics[i] = resolvedTypeArgument;
 						}
 						else {
-							generics[i] = ResolvableType.forType(typeArgument).resolve();
+							generics[i] = ResolvableType.forType(typeArgument);
 						}
 					}
+					else if (typeArgument instanceof ParameterizedType) {
+						generics[i] = ResolvableType.forType(resolveType(typeArgument, contextClass));
+					}
 					else if (typeArgument instanceof WildcardType wildcardType) {
-						generics[i] = resolveWildcard(wildcardType, contextType).resolve();
+						ResolvableType result = resolveWildcard(wildcardType, contextType);
+						if (result.resolve() == null) {
```

Comment:
I wonder if that logic couldn't be applied in `resolveWildcard` since that's the only use of it. This would allow this to be a one liner like the other cases.

Reply:
> I wonder if that logic couldn't be applied in `resolveWildcard` since that's the only use of it. This would allow this to be a one liner like the other cases.

This change makes `GenericTypeResolverTests.resolveWildcardTypeWithUnbounded()` pass.

Reply:
OK I am not saying removing. I am thinking about moving `result.resolve()` to the `resolveWildcard` method.

Reply:
OK, updated as you suggested.